financial-agent-py/agent/index_nevermined.py

```diff
@@ -0,0 +1,231 @@
+
+When discussing investments:
+Focus on general principles and educational concepts while explaining both potential benefits and risks in a conversational way. Naturally emphasize the importance of diversification and long-term thinking. Gently remind users to only invest what they can afford to lose and suggest they research thoroughly while considering their personal financial situation. Make these important points feel like natural parts of the conversation rather than formal warnings."""
+
+# Calculate dynamic credit amount based on token usage
+def calculate_credit_amount(tokens_used: int, max_tokens: int) -> int:
+    """
+    Formula: 10 * (actual_tokens / max_tokens)
+    This rewards shorter responses with lower costs
+    """
+    token_utilization = min(tokens_used / max_tokens, 1.0)  # Cap at 1
+    base_credit_amount = 10 * token_utilization
+    credit_amount = max(math.ceil(base_credit_amount), 1)  # Minimum 1 credit
+
+    print(f"Token usage: {tokens_used}/{max_tokens} ({token_utilization * 100:.1f}%) - Credits: {credit_amount}")
+
+    return credit_amount
+
+# Store conversation history for each session
+sessions: Dict[str, List[Dict[str, str]]] = {}
+
+# Request/Response models
+class AskRequest(BaseModel):
+    input_query: str
+    sessionId: Optional[str] = None
+
+class RedemptionResult(BaseModel):
+    creditsRedeemed: int
+    error: Optional[str] = None
+
+class AskResponse(BaseModel):
+    output: str
+    sessionId: str
+    redemptionResult: Optional[RedemptionResult] = None
+
+# Handle financial advice requests with Nevermined payment protection and observability
+@app.post("/ask", response_model=AskResponse)
+async def ask_financial_advice(request: AskRequest, authorization: Optional[str] = Header(None)):
+    try:
+        # Extract authorization details from request headers
+        auth_header = authorization or ""
+        requested_url = f"{NVM_AGENT_HOST}/ask"
+        http_verb = "POST"
+
+        # Check if user is authorized and has sufficient balance
+        agent_request = payments.requests.start_processing_request(
+            NVM_AGENT_ID,
+            auth_header,
+            requested_url,
+            http_verb
+        )
+
+        # Reject request if user doesn't have credits or subscription
+        balance_info = agent_request.get("balance", {})
+        is_subscriber = balance_info.get("isSubscriber", False)
+        balance_raw = balance_info.get("balance", 0)
+        # Convert balance to int if it's a string
+        if isinstance(balance_raw, str):
+            balance_amount = int(balance_raw) if balance_raw.isdigit() else 0
+        else:
+            balance_amount = balance_raw
+        if not is_subscriber or balance_amount < 1:
+            raise HTTPException(status_code=402, detail="Payment Required")
+
+        # Extract access token for credit redemption
+        request_access_token = auth_header.replace("Bearer ", "", 1) if auth_header.startswith("Bearer ") else auth_header
```

Comment:
I don't think this is a problem with this pr in specific but I don't like the inconsistency of how we handle the auth_header. It either enforce it to be a bearer token or we don't.